File: graph-master/number_of_islands.py
class Solution:
    """
    @param grid: a boolean 2D matrix
    @return: an integer
    """

    def numIslands(self, grid):
        # write your code here
        # use bfs to search the grid

        if len(grid) == 0:
            return 0

        delat = [[0, 1], [1, 0], [0, -1], [-1, 0]]
        row, column = len(grid), len(grid[0])
        result = 0

        for i in range(row):
            for j in range(column):
                if grid[i][j] == 1:
                    self.bfs(grid, i, j, delat)
                    result += 1

        return result

    def bfs(self, grid, i, j, delat):
        queue = []
        queue.append((i, j))
        grid[i][j] = 0
        while len(queue) > 0:
            x, y = queue.pop(0)
            for dx, dy in delat:
                if 0 <= dx + x < len(grid) and 0 <= dy + y < len(grid[0]) and grid[dx + x][dy + y] == 1:
                    grid[dx + x][dy + y] = 0
                    queue.append((dx + x, dy + y))

File: src/depth-first-searches/number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        if(grid.length == 0 || grid[0].length == 0)
            return 0;
        int m = grid.length;
        int n = grid[0].length;
        int result = 0;
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(grid[i][j] == '1'){
                    dfs(grid, i, j, m, n);
                    result += 1;
                }
            }
        }
        
        return result;
    }
    
    public void dfs(char[][] grid, int i, int j, int m, int n){
        if(i < 0 || i >= m || j < 0 || j >= n || grid[i][j] != '1')
            return;
        grid[i][j] = '0';
        dfs(grid, i + 1, j, m, n);
        dfs(grid, i - 1, j, m, n);
        dfs(grid, i, j + 1, m, n);
        dfs(grid, i, j - 1, m, n);
    }
    
}